_pytest/assertion/rewrite.py

```diff
@@ -258,22 +260,21 @@ def _write_pyc(state, co, source_stat, pyc):
     # sometime to be able to use imp.load_compiled to load them. (See
     # the comment in load_module above.)
     try:
-        fp = open(pyc, "wb")
-    except IOError:
-        err = sys.exc_info()[1].errno
-        state.trace("error writing pyc file at %s: errno=%s" % (pyc, err))
+        with atomicwrites.atomic_write(pyc, mode="wb", overwrite=True) as fp:
+            fp.write(imp.get_magic())
+            mtime = int(source_stat.mtime)
+            size = source_stat.size & 0xFFFFFFFF
+            fp.write(struct.pack("<ll", mtime, size))
+            if six.PY2:
+                marshal.dump(co, fp.file)
```

Comment:
that bit looks strange why is that kind of access needed?

Reply:
It fails on PY2 because `fp` is not a real file object, it is a `_TemporaryFile` (or something like that). In PY3 just passing `fp` to `marshal.dump` works.

It is probably possible to use `marshal.dump(co, fp.file)` in both Python 2 and 3, but I prefer to separate those cases so when the time comes around to drop Python 2 it is clear that we can keep the "cleaner" version.